crates/astria-sequencer/src/accounts/ics20_withdrawal.rs

```diff
@@ -0,0 +1,132 @@
+    },
+    transaction::action_handler::ActionHandler,
+};
+
+fn withdrawal_to_unchecked_ibc_packet(withdrawal: &Ics20Withdrawal) -> IBCPacket<Unchecked> {
+    IBCPacket::new(
+        PortId::transfer(),
+        withdrawal.source_channel.clone(),
+        withdrawal.timeout_height,
+        withdrawal.timeout_time,
+        withdrawal.packet_data(),
+    )
+}
+
+#[async_trait::async_trait]
+impl ActionHandler for Ics20Withdrawal {
+    #[instrument(skip(self))]
+    async fn check_stateless(&self) -> Result<()> {
+        if self.timeout_time == 0 {
+            anyhow::bail!("timeout time must be non-zero");
+        }
+
+        // NOTE (from penumbra): we could validate the destination chain address as bech32 to
+        // prevent mistyped addresses, but this would preclude sending to chains that don't
+        // use bech32 addresses.
+        Ok(())
+    }
+
+    #[instrument(skip(self, state))]
+    async fn check_stateful<S: StateReadExt + 'static>(
+        &self,
+        state: &S,
+        from: Address,
+        _fee_asset_id: asset::Id,
+    ) -> Result<()> {
+        let packet: IBCPacket<Unchecked> = withdrawal_to_unchecked_ibc_packet(self);
+        state
+            .send_packet_check(packet)
+            .await
+            .context("packet failed send check")?;
+
+        let from_transfer_balance = state
+            .get_account_balance(from, self.denom.id())
+            .await
+            .context("failed getting `from` account balance for transfer")?;
+        ensure!(
+            from_transfer_balance > self.amount,
+            "insufficient funds for transfer"
+        );
+
+        Ok(())
+    }
+
+    #[instrument(skip(self, state))]
+    async fn execute<S: StateWriteExt>(
+        &self,
+        state: &mut S,
+        from: Address,
+        _fee_asset_id: asset::Id,
+    ) -> Result<()> {
+        let checked_packet = withdrawal_to_unchecked_ibc_packet(self).assume_checked();
+
+        let from_transfer_balance = state
+            .get_account_balance(from, self.denom.id())
+            .await
+            .context("failed getting `from` account balance for transfer")?;
+        ensure!(
+            from_transfer_balance > self.amount,
+            "insufficient funds for transfer"
+        );
+
+        state
+            .put_account_balance(from, self.denom.id(), from_transfer_balance - self.amount)
+            .context("failed to update sender balance")?;
+
+        // if we're the source, move tokens to the escrow account,
+        // otherwise the tokens are just burned
+        if is_source(
+            checked_packet.source_port(),
+            checked_packet.source_channel(),
+            &self.denom,
+        ) {
+            let channel_balance = state
+                .get_ibc_channel_balance(&self.source_channel, self.denom.id())
+                .await
+                .context("failed to get channel balance")?;
+
+            state
+                .put_ibc_channel_balance(
```

Comment:
Do we need to worry about making the entirety of these steps a transaction similar to how SQL transactions are atomic (if one fails, all fail or are rolled back)?

Reply:
mentioned this before, but check `deliver_tx`, it starts an atomic tx for the db that's only committed if the tx succeeds

Reply:
OK, cool. I have to go over how new entries in the DB are created to understand the flow again.